zellij-server/src/logging_pipe.rs

```diff
@@ -0,0 +1,258 @@
+use std::{
+    collections::VecDeque,
+    io::{Read, Seek, Write},
+};
+
+use log::{debug, error, info};
+use wasmer_wasi::{WasiFile, WasiFsError};
+use zellij_utils::serde;
+
+use chrono::prelude::*;
+use serde::{Deserialize, Serialize};
+
+// 16kB log buffer
+const ZELLIJ_MAX_PIPE_BUFFER_SIZE: usize = 16_384;
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(crate = "self::serde")]
+pub struct LoggingPipe {
+    buffer: VecDeque<u8>,
+    plugin_name: String,
+    plugin_id: u32,
+}
+
+impl LoggingPipe {
+    pub fn new(plugin_name: &str, plugin_id: u32) -> LoggingPipe {
+        info!(
+            "|{:<25.25}| {} [{:<10.15}] Creating decorating pipe for plugin: {}!",
+            plugin_name,
+            Local::now().format("%Y-%m-%d %H:%M:%S.%3f"),
+            format!("id: {}", plugin_id),
+            plugin_name
+        );
+        LoggingPipe {
+            buffer: VecDeque::new(),
+            plugin_name: String::from(plugin_name),
+            plugin_id,
+        }
+    }
+
+    fn log_message(&self, message: &str) {
+        debug!(
+            "|{:<25.25}| {} [{:<10.15}] {}",
+            self.plugin_name,
+            Local::now().format("%Y-%m-%d %H:%M:%S.%3f"),
+            format!("id: {}", self.plugin_id),
+            message
+        );
+    }
+}
+
+impl Read for LoggingPipe {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        let amt = std::cmp::min(buf.len(), self.buffer.len());
```

Comment:
Usually when you run into manual indexing like this in Rust there is another way to go about it. Sometimes that other way can be super elegant, sometimes less so... I think this is one of the less elegant times, but here is a slightly more "functional" (as in paradigm) approach if you are interested:

![image](https://user-images.githubusercontent.com/6251883/125829925-a65b8728-81f1-4274-9b67-1bb82488fd09.png)

The same for copy-paste:
```rs
impl Read for LoggingPipe {
    fn read(&mut self, mut buf: &mut [u8]) -> std::io::Result<usize> {
        let data: Vec<_> = self.buffer.drain(..buf.len()).collect();
        buf.write(&data)
    }
}
```

Reply:
Thanks, I like your code more. To be honest, this was straight copied from wasmer-wasi `Pipe`.
Thinking about it, I think we should remove this code altogether. We're draining our buffer in `flush()`, so I guess if someone were to chain this `LoggingPipe` with something that would read from it, it would see very weird behavior and also logs would be screwed up.

I changed to your version, added note that it should probably be removed.

One caveat:
I had to do:
`buf.as_mut().write(&data)`
without `as_mut()` I was getting:
```

error[E0596]: cannot borrow `buf` as mutable, as it is not declared as mutable
  --> zellij-server/src/logging_pipe.rs:47:9
   |
47 |         buf.write(&data)
   |         ^^^
   |         |
   |         cannot borrow as mutable
   |         try removing `&mut` here

error: aborting due to previous error; 1 warning emitted

```

I have no idea that's the case. I thought `buf` is already mut?

Reply:
I think Rust is being a tad strange here. When calling `.write()` it's attempting to take a `&mut` of the `buf`. Even if the `buf` *stores* a mutable reference, the `buf` variable itself is non mutable, so Rust doesn't let you take a mutable reference of it. If you add `mut buf: ...` instead of just `buf: ...` in the arguments of `read()`, that should fix things (I think)!

With that being said, you seem write about us not using this! I think that, since people really shouldn't be calling `.read()`, we shouldn't let them! How do you feel about mirroring the `Seek` implementation with something like:
![image](https://user-images.githubusercontent.com/6251883/126060407-73ee1734-0ae9-4d7e-9529-ed5b2b736863.png)

Copy paste:
```rs
fn read(&mut self, _buf: &mut [u8]) -> std::io::Result<usize> {
    Err(std::io::Error::new(
        std::io::ErrorKind::Other,
        "can not read a LoggingPipe",
    ))
}
```

Reply:
Yeah, it's good idea to just return error for reads. Obviously, it can be changed in future by not draining buffer but keeping a pointer instead. Changed it to return error, should be fine for now.